### PR TITLE
Add documentation for using expressions as service factories

### PR DESCRIPTION
**Pull request by jvasseur**
Doc PR for https://github.com/symfony/symfony/pull/45512

### CODE REVIEW

**Comment by javiereguiluz** [comment]
Closing because the related code PR (https://github.com/symfony/symfony/pull/45447) was closed without merging.

**Comment by jvasseur** [comment]
The code PR was replaced by https://github.com/symfony/symfony/pull/45512 which is still opened so this PR could still be relevant.

**Comment by nicolas-grekas** [comment]
Code PR has been merged, time to finish this one!

File: service_container/expression_language.rst [hunk 1]
```diff
@@ -125,3 +125,44 @@ via a ``container`` variable. Here's another example:
 
 Expressions can be used in ``arguments``, ``properties``, as arguments with
 ``configurator`` and as arguments to ``calls`` (method calls).
+
+You can also use expressions as factories to create services based on an expression:
```
Comment (nicolas-grekas) [inline service_container/expression_language.rst]: What about shortening this to `You can also use expressions as service factories:`

**Comment by nicolas-grekas** [comment]
We should also tell about the args/arg() var/function btw.

**Comment by jvasseur** [comment]
I've update the PR to mention the `arg` function and add the `versionadded` block.

File: service_container/expression_language.rst [hunk 2]
```diff
@@ -125,3 +125,50 @@ via a ``container`` variable. Here's another example:
...
+        use App\Mailer;
+
+        return function(ContainerConfigurator $configurator) {
+            $services = $configurator->services();
+
+            $services->set(Mailer::class)
+                ->factory(expr("parameter('some_param') ? service('some_service') : service('some_other_service')"));
+        };
+
+In this context, you have access to the `arg` function that allows getting the value of arguments passed to the factory.
```
Comment (nicolas-grekas) [inline service_container/expression_language.rst]: Worth a small example, isn't it?

**Review by nicolas-grekas — APPROVED** [review]

File: service_container/expression_language.rst [hunk 3]
```diff
@@ -125,3 +125,54 @@ via a ``container`` variable. Here's another example:
...
+You can also use expressions as service factories:
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        # config/services.yaml
+        services:
+            App\Mailer:
+                factory: "@=parameter('some_param') ? service('some_service') : arg(0)"
```
Comment (xabbuh) [inline service_container/expression_language.rst:138]: Does this require the class referenced by the `some_service` to have a `__invoke()` method?

File: service_container/expression_language.rst [hunk 4]
```diff
@@ -125,3 +125,54 @@ via a ``container`` variable. Here's another example:
...
+
+        return function(ContainerConfigurator $configurator) {
+            $services = $configurator->services();
+
+            $services->set(Mailer::class)
+                ->factory(expr("parameter('some_param') ? service('some_service') : arg(0)"))
+                ->args([service('some_other_service')]);
+        };
+
+In this context, you have access to the `arg` function that allows getting the value of arguments passed to the factory.
```
Comment (xabbuh) [inline service_container/expression_language.rst]:
```suggestion
In this context, you have access to the ``arg`` function that allows getting the value of arguments passed to the factory.
```

File: service_container/expression_language.rst (hunk 3 above)
Comment (jvasseur) [reply service_container/expression_language.rst:138]: No, it doesn't use the result of the expression as of factory but the expression itself is the factory. In this case this means hat the `App\Mailer` service will be an alias to `some_service` if the parameter `some_param` is `true`.

**Comment by wouterj** [comment]
Thank you @jvasseur! It's great to retrieve a doc PR for a new code feature :)

Fyi, after the merge I've moved the section to the factories article (and changed the example a bit) in 154d66768

**Comment by jvasseur** [comment]
@wouterj thanks for merging, the updated examples looks really good with the associated explanations.

**Comment by wkania** [comment]
@wouterj https://github.com/symfony/symfony-docs/commit/154d6676863702c09c3d2c92bcd4a03aa58f4d93 introduced a bug
```
Found invalid reference "/service_container/expressions" in file       
"service_container/factories"  
```
Job "Build the docs" is now broken.

**Comment by xabbuh** [comment]
FTR, that was fixed in da915703a7222b68378be3b9375fa0d7b1bae02c.

**Pull request merged**